HW-3 MVP 적용

chuuuul opened the pull request:
- Model  ↔ Present ↔ View의 구조
- View는 UI의 I/O를 Present에 전달한다.
- Present는 로직을 처리한다.
- Contract는 View와 Present의 인터페이스를 구성한다.
- 각 액티비티 프래그먼트 마다 각각의 구조를 적용시킨다.

chuuuul committed:
HW-3 리팩토링 : SearchContract Interface 생성

chuuuul committed:
HW-3 리펙토링 : SearchPresenter Interface 생성

chuuuul committed:
HW-3 리팩토링 : 필요없는 주석 제거

chuuuul committed:
HW-3 리팩토링 : Adapter의 역할  ViewHolder의 onBind로 분리

- onBind에서 각각의 view에 text등을 수정한다.

chuuuul committed:
HW-3 리팩토링 : 클래스명 MovieListAdapter -> MovieAdapter로 변경


- 기능(feat): 기능 추가, 삭제, 변경 (제품 코드 수정 발생)
- 버그(fix): 버그 수정 (제품 코드 수정 발생)
- 리팩토링(refactor): 코드 리팩토링 (제품 코드 수정 발생)
- 형식(style): 코드 형식, 정렬 등의 변경 (제품 코드 수정 발생. 하지만 동작에 영향을 주는 변경은 없음)
- 테스트(test): 테스트 코드 추가, 삭제, 변경 등 (제품 코드 수정 없음. 테스트 코드에 관련된 모든 변경에 해당)
- 문서(docs): 문서 추가, 삭제, 변경 (코드 수정 없음)
- 기타: 위에 해당되지 않는 모든 변경(예: 빌드 스크립트 수정, 패키지 배포 설정 변경 등)을 포함 (코드 수정 없음)

chuuuul committed:
HW-3 리팩토링 : MovieAdapterContract 생성

chuuuul committed:
HW-3 리팩토링 : MovieAdapterPresenter 생성

chuuuul committed:
HW-3 리팩토링 : MovieHolderContract 생성

chuuuul committed:
HW-3 리팩토링 : MovieHolderPresenter 생성

chuuuul committed:
HW-3 리팩토링 : Movie Adapter에서 MovieViewHolder 분리

 - 코드 양이 많아짐에 따라 분리하여 관리



- 기능(feat): 기능 추가, 삭제, 변경 (제품 코드 수정 발생)
- 버그(fix): 버그 수정 (제품 코드 수정 발생)
- 리팩토링(refactor): 코드 리팩토링 (제품 코드 수정 발생)
- 형식(style): 코드 형식, 정렬 등의 변경 (제품 코드 수정 발생. 하지만 동작에 영향을 주는 변경은 없음)
- 테스트(test): 테스트 코드 추가, 삭제, 변경 등 (제품 코드 수정 없음. 테스트 코드에 관련된 모든 변경에 해당)
- 문서(docs): 문서 추가, 삭제, 변경 (코드 수정 없음)
- 기타: 위에 해당되지 않는 모든 변경(예: 빌드 스크립트 수정, 패키지 배포 설정 변경 등)을 포함 (코드 수정 없음)

chuuuul committed:
HW-3 리팩토링 : MovieViewHolder MVP 패턴 적용

chuuuul committed:
HW-3 리팩토링 : MovieAdapter에 MVP 적용

chuuuul committed:
HW-3 리팩토링 : 변수명 변경

 - naverRemoteDataSourceImpl -> naverRemoteDataSource
 - naverLocalDataSourceImpl -> naverLocalDataSourceImpl

chuuuul committed:
HW-3 리펙토링 : 변수명 변경

 - edt_search_searchName -> edt_search_keyword

chuuuul committed:
HW-3 리팩토링 : SearchActivity MVP로 변경

developer-sungjae commented:
안녕하세요~

chuuuul committed:
HW-3 리팩토링 : 메서드 이름 변경
 - View의 메서드중 set- -> show- 으로 view의 느낌으로 변경

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/adapter/MovieHolderPresenter.kt:
```diff
@@ -0,0 +1,23 @@
+import com.example.architecture.util.ConstValue
+
+class MovieHolderPresenter(
+    private val view: MovieHolderContract.View
```
안쓰는 것 같은데 확인 부탁드립니다.

chuuuul replied:
84ec814
감사합니다. 수정했습니다.
onBind 작업을 ViewHolder에서 하지 않고 Presenter에서 작업하도록 하여 view로 전달하도록 하였습니다.

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/adapter/MovieHolderPresenter.kt:
```diff
@@ -0,0 +1,23 @@
+) : MovieHolderContract.Presenter {
+
+    override fun removeMarkupTag(html: String): String {
+        return html.replace("<b>", "").replace("</b>", "")
+    }
+
```
검색어 '안'을 쳐보시면 html 파싱이 안되는 것을 볼 수가 있습니다. 이 부분에 대해서도 파싱이 잘 되도록 변경 해주세요!

chuuuul replied:
a96871c
감사합니다. 성재님 코드 참고하여 수정했습니다!

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/SearchActivity.kt:
```diff
@@ -1,88 +1,57 @@
     }
 
-    // * RecyclerView
-
-    private fun setRecyclerview() {
-//        adapter.addNewItems(movieList)
-        recyclerview_search_movieList.adapter = adapter
+    override fun showToast(@StringRes resId: Int) {
+        Toast.makeText(this, getText(resId), Toast.LENGTH_SHORT).show()
```
어떤 토스트인지 메서드만으로는 추측할 수 없을 것 같습니다.

chuuuul replied:
12d0368
감사합니다. 수정했습니다.

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/adapter/MovieAdapter.kt:
```diff
@@ -0,0 +1,49 @@
+        context.startActivity(intent)
+    }
+
+    override fun getItemCount(): Int {
+        return movieAdapterPresenter.getMovieCount()
+    }
```
어떤부분은 괄호가 있고 어떤 부분은 없는데 통일 해주시면 좋을 것 같습니다.

chuuuul replied:
b85c385
감사합니다. 수정했습니다

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/adapter/MovieHolderPresenter.kt:
```diff
@@ -0,0 +1,23 @@
+
+    override fun checkImageUrl(imageUrl: String): String {
+        return if (imageUrl.isBlank()) {
+            ConstValue.NO_IMAGE_URL
```
AppUtil은 import하셨는데 ConstValue는 따로 남겨 놓아서 이것도 없애는게 좋지 않을까요?

chuuuul replied:
f352da6
감사합니다. 수정했습니다.

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/SearchActivity.kt:
```diff
@@ -1,88 +1,57 @@
-        if (isValidKeyword(keyWord)) {
-            naverRepositoryImpl.getMovieList(
-                keyWord,
-                onSuccess = this::onSuccess,
-                onFailure = this::onFailure
-            )
+        btn_search_searchButton.setOnClickListener {
+            searchPresenter.searchMovie(edt_search_keyword.text.toString())
```
가독성을 위해서 매우 긴 parameter는
`
val a = '매우 긴 value'
`
`
method(a)
`
식으로 쓰는게 가독성에 훨씬 더 보기 좋은 것 같습니다

chuuuul replied:
a96871c
감사합니다. 수정했습니다.

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/SearchPresenter.kt:
```diff
@@ -0,0 +1,55 @@
+        Log.d("chul", "OnFailure : $t")
+    }
+
+    override fun searchMovie(actionId: Int, keyword: String): Boolean {
+        return if (actionId == EditorInfo.IME_ACTION_SEARCH) {
+            searchMovie(keyword)
+            true
+        } else {
+            false
+        }
+    }
+
+    override fun isValidKeyword(keyword: String): Boolean {
+
+        if (keyword.isBlank()) {
+            view.showToast(R.string.empty_keyword)
+            return false
+        }
+        return true
+    }
```
어디는 return if이고 어디는 if문이고 어디는 if문 + 밑의 return을 붙여주셨는데 혹시 가독성을 생각하시고 정하신건가요~? 아니면 기준이 있으신지 궁금합니다. 
기준이 없으시다면 통일하면 좋을 것 같습니다.

chuuuul replied:
ae45975
감사합니다. 수정했습니다!

developer-sungjae commented:
그 외에는 저보다 더 자세하게 구현하셨는데 딱히 코멘트 드릴게 없네요.

developer-sungjae commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/repository/NaverRepositoryImpl.kt:
```diff
@@ -18,9 +20,9 @@ class NaverRepositoryImpl(context : Context) : NaverRepository {
     ) {
 
-        naverLocalDataSourceImpl.getMovieList(keyword, onSuccess)
+        naverLocalDataSource.getMovieList(keyword, onSuccess)
```
혹시 로컬무비리스트 가져오는 동작방식은 어떻게 되나요~?

chuuuul committed:
HW-3 리팩토링 : view의 onBind()의 작업을 Present로 이동

chuuuul committed:
HW-3 형식 : 주석 추가

 - "Keyboard의 엔터로 검색 작업 주석" 추가

chuuuul committed:
HW-3 리팩토링 : showToast를 세분화하여 작성

- showMessageEmptyKeyword()
- showMessageEmptyResult()

chuuuul committed:
HW-3 리팩토링 : 1줄짜리 함수 브라켓 통일

chuuuul committed:
HW-3 리팩토링 : import 수정

chuuuul committed:
HW-3 리팩토링 : 긴 인자 값 변수로 담아서 사용

chuuuul committed:
HW-3 리팩토링 : return 형식 통일

chuuuul committed:
HW-3 수정 : html 태그 제거 버그 수정
 - 네이버 영화 api에서 오는 -에 대한 유니코드를 제거 못하는 버그 발생
 - Apache에서 제공하는 html 라이브러리와 jsoup (html parse)라이브러리를 통해 해결

developer-sungjae commented:
고생하셨습니다.

sjjeong commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/SearchPresenter.kt:
```diff
@@ -0,0 +1,57 @@
+
+        return if (keyword.isBlank()) {
+            view.showMessageEmptyKeyword()
+            return false
```
return 제거 해주세요

chuuuul replied:
b48fd5d
감사합니다. 수정했습니다!

sjjeong commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/adapter/MovieAdapterContract.kt:
```diff
@@ -0,0 +1,20 @@
+
+    interface Presenter {
+
+        fun getMovie(position: Int): MovieModel
+        fun getMovieCount(): Int
```
mvp에서 Cntract에 있는 함수들은 return이 없는 형태로 만들어 주셔야 해요
다른 부분도 찾아서 같이 수정해주세요

chuuuul replied:
97839a2
감사합니다. 수정했습니다.

chuuuul committed:
HW-3 리팩토링 : Contract에 선언된 메서드의 return 타입 제거

chuuuul committed:
HW-3 리팩토링 : Search의 Contract에 선언된 메서드의 return 타입 제거

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
